### ISSUE TITLE
Buffering when playing

### ISSUE THREAD

**Issue by Gythion** [anxdpanic/plugin.video.youtube#211]
### Context

Please provide any relevant information about your setup

* Add-on Version: 6.8.17 matrix
* Kodi Version: 19.0
* Kodi GUI Language: Ita
* Operating System: Win10
* Operating System Language: Ita

------

### Expected Behavior

Normally playing without buffering

------

### Current Behavior

Buffering at almost every played video in FHD or HD

### Log

t.ly/9ItW

------

### Additional Information

How about? It is already several months that there is this problem. I use the official YT app in FireStick, and no buffering.

------

**Comment by jonisb** [comment]
Update the plugin to [6.8.18.alpha1](https://github.com/anxdpanic/plugin.video.youtube/releases/tag/6.x.x-dev) as it seems to work much better with the buffering
#163 issue is related to this

**Comment by Rialtho-the-Magnificent** [comment]
I have the same. More and more videos show this behaviour. Have tried to reduce resolution, but even with 480 setting I still get buffering. Also, when stopping the video and restarting it, it takes several attempts before it finally decides to play normal. Some videos however take more than 5 times and that is too much for me. The logs show nothing that might explain what is going on.

**Comment by probonopd** [comment]
Same here. The issue seems to persist in 6.8.18.alpha1, at least with some videos.
Are we getting throttled?

**Comment by terxw** [comment]
Buffering also in  6.8.18.alpha1
Edit: 6.8.18.alpha2 is working OK, no bufering

**Comment by ghost** [comment]
Happens in Linux too. Not always, occasionally.

**Comment by Rialtho-the-Magnificent** [comment]
I use it on a PI 3b with OSMC but the buffering problem is getting worse and worser. Started couple of weeks ago. I guess it is something that Google did and it is related to the amount of ads that are suppose to be shown in front of the selected videos, especially when the video is new there is a time period (which varies per channel) where ads are non-skippable if viewed via other means, like the YT app on a tablet or phone. Older videos have this problem also, but one or two times stopping and starting the video solves the problem.

**Comment by arncor** [comment]
> I use it on a PI 3b with OSMC but the buffering problem is getting worse and worser. Started couple of weeks ago. I guess it is something that Google did and it is related to the amount of ads that are suppose to be shown in front of the selected videos, especially when the video is new there is a time period (which varies per channel) where ads are non-skippable if viewed via other means, like the YT app on a tablet or phone. Older videos have this problem also, but one or two times stopping and starting the video solves the problem.

interesting comment, but I haven't got any buffering issues when I watch YouTube using the Android Newpipe app (which bans ads) or on the computer using Firefox + uBlock Origin (which also bans ads). It seems to be really specific to Kodi/YouTube_addon/InputStream_adaptative.

PS : I also got buffering issues on Raspberry Pi 4, Kodi 18.7, YouTube 6.8.17, InputStream Adaptative 2.4.4.
When buffering, I noticed the network download speed settles around 70-90KB/s (out of a 300Mbps connexion).
When NOT buffering, the network usage is different, with occasionnal spikes of hundreds of KB/s and 0KB/s in between.

Also the YouTube addon cannot allow more than 100MB (default is 70MB) of cache, which I think is pretty low.
Cache settings in advancedsettings.xml file such as buffermode, memorysize and readfactor doesn't seem to apply to video played with the youtbe addon.

PS2 : I mostly play Youtube videos on Kodi using my phone+Newpipe > "play with kodi" option. So the URL sent to kodi is done via newpipe, maybe it has an influence...

**Comment by Rialtho-the-Magnificent** [comment]
Thanks for the additional info. I also have no problems running adfree YT on my PC (using an modified etc\hosts file). But the buffering seems to be sort of random. Yesterday, I also had a 403 error (according to the log file of kodi) on a number of YT videos that are mainly music. Not sure if this is related. And then there are the videos that give the sad face when selected. That usually goes away after several hours. I still believe that has something to do with the non-skippable adds that Google wants to serve as hours later the video does play normal (or also has buffering issues, and you have to retry several times). (all using 6.8.18...)

**Comment by pitsi** [comment]
If @anxdpanic is reading this, can you please add the development version of the addon in your testing repo? It will help us who use the repo get the newest releases and do the testing early. Right now it hosts 6.8.16, which is older than the one in the main kodi repo.
Thank you in advance.

**Comment by probonopd** [comment]
Still getting the buffering issue with `plugin.video.youtube-6.8.18.alpha2.zip`, e.g., on this video:
https://www.youtube.com/watch?v=a3oQy6v3MrA

**Comment by probonopd** [comment]
Here is a workaround that works reliably around this issue (while waiting for a proper fix):

1. Install and configure https://github.com/lekma/plugin.video.invidious/ (use release 0.5.7 if you are on a Python 2 based system). Configure it to use `vid.puffyan.us` as the instance
2. In `plugin.video.youtube`, add a "Play with Invidious" context menu:

At the end of `plugin.video.youtube/resources/lib/youtube_plugin/youtube/helper/yt_context_menu.py`, add:

```
def append_play_with_invidious(context_menu, provider, context, video_id):
    context_menu.append(("Play with Invidious", 'PlayMedia(plugin://plugin.video.invidious/?action=video&videoId=%s)' % video_id))
```

In `plugin.video.youtube/resources/lib/youtube_plugin/youtube/helper/utils.py`, add directly above the line `def update_play_info(provider, context, video_id, video_item, video_stream, use_play_data=True):`

```
        yt_context_menu.append_play_with_invidious(context_menu, provider, context, video_id)
```

Restart Kodi (or deactivate and activate the YouTube addon) for the changes to take effect.

Now select "Play with Invidious" from the context menu to enjoy stutter-free instant playback.

Does anyone know how to make "Play with Invidious" the default action when a video is clicked?

---

I haven't investigated how Invidious generates the information that ends up in

* https://vid.puffyan.us/api/manifest/dash/id/aqz-KE-bpKQ
* https://vid.puffyan.us/api/v1/videos/aqz-KE-bpKQ?&pretty=1

but it seems it is golden.

**Comment by tssajo** [comment]
@probonopd Unfortunately, the workaround you posted doesn't work for me.
In fact, Invidious cannot play a single video on my Raspberry Pi 3 Model B Plus running LibreELEC.
I managed to install plugin.video.invidious to v0.6.6 from Lekma's repository, from here:
https://github.com/lekma/repository.lekma/

Here is what gets logged:
```
2021-10-10 23:36:57.971 T:1235366784  NOTICE: [plugin.video.invidious] request.url: https://invidious.snopyta.org/api/v1/videos/tflC9c4KhFo
2021-10-10 23:37:00.642 T:1147577216  NOTICE: [plugin.video.invidious] video.url: https://invidious.snopyta.org/api/manifest/dash/id/tflC9c4KhFo
2021-10-10 23:37:01.582 T:1937289280  NOTICE: VideoPlayer::OpenFile: plugin://plugin.video.invidious/?action=video&videoId=tflC9c4KhFo
2021-10-10 23:37:01.590 T:1218581376  NOTICE: Creating InputStream
2021-10-10 23:37:01.677 T:1218581376   ERROR: CCurlFile::FillBuffer - Failed: HTTP returned error 403
2021-10-10 23:37:01.677 T:1218581376   ERROR: CCurlFile::Open failed with code 403 for https://invidious.snopyta.org/api/manifest/dash/id/tflC9c4KhFo:
2021-10-10 23:37:01.677 T:1218581376   ERROR: AddOnLog: InputStream Adaptive: Cannot download https://invidious.snopyta.org/api/manifest/dash/id/tflC9c4KhFo
2021-10-10 23:37:01.677 T:1218581376   ERROR: AddOnLog: InputStream Adaptive: Could not open / parse mpdURL (https://invidious.snopyta.org/api/manifest/dash/id/tflC9c4KhFo)
2021-10-10 23:37:01.678 T:1218581376   ERROR: CVideoPlayer::OpenInputStream - error opening [plugin://plugin.video.invidious/?action=video&videoId=tflC9c4KhFo]
2021-10-10 23:37:01.678 T:1218581376  NOTICE: CVideoPlayer::OnExit()
2021-10-10 23:37:01.794 T:1937289280  NOTICE: CVideoPlayer::CloseFile()
2021-10-10 23:37:01.794 T:1937289280  NOTICE: VideoPlayer: waiting for threads to exit
2021-10-10 23:37:01.794 T:1937289280  NOTICE: VideoPlayer: finished waiting
```
As you can see, Invidious receives 403 HTTP errors. But why?
Does anyone have any idea on how to solve this?

**Comment by probonopd** [comment]
Try `vid.puffyan.us` as the instance instead of `invidious.snopyta.org`. snopyta.org seemingly does not allow API access.

**Comment by tssajo** [comment]
@probonopd Thanks for the tip! But I am in Europe. Anyhow, I just checked the URL with my PC and it downloaded the file just fine, this one: https://invidious.snopyta.org/api/manifest/dash/id/tflC9c4KhFo
So I didn't get a 403 error on my PC. I experiment with Curl on my Raspberry Pi. If I find a solution I will post it here for others.

**Comment by tssajo** [comment]
OK, Invidious finally works. Both `vid.puffyan.us` and `invidious.namazso.eu` works fine. I picked the latter one from this list `https://redirect.invidious.io/watch?v=dMRH96Ew5G4`

**Comment by bzarky** [comment]
I am having the same playback issues as well and it's worse  after the 19.2 kodi update. 

Here's my debug log trying to play a youtube video. 
System is Ubuntu 21.04 on am64

[kodi.log](https://github.com/anxdpanic/plugin.video.youtube/files/7318635/kodi.log)

**Comment by RNavega** [comment]
> Does anyone know how to make "Play with Invidious" the default action when a video is clicked?

@probonopd 

1. When a video item is created, it uses the `/play/video_id=...` route, created in here:  
https://github.com/anxdpanic/plugin.video.youtube/blob/36e037518bbb14a06a4928c9dd5b77baaa70c3d3/resources/lib/youtube_plugin/youtube/helper/v3.py#L54

2. That route is processed in here:  
https://github.com/anxdpanic/plugin.video.youtube/blob/36e037518bbb14a06a4928c9dd5b77baaa70c3d3/resources/lib/youtube_plugin/youtube/provider.py#L683-L698

So if you want to have "Play With Invidious" as the default action, while still supporting playing in the original way, then you'll have to change that line in item 1 to be something like:
```
            item_uri = context.create_uri(['playInvidious'], item_params)
```

And then add this new "playInvidious" route to your [provider.py](https://github.com/anxdpanic/plugin.video.youtube/blob/master/resources/lib/youtube_plugin/youtube/provider.py):
```
        @kodion.RegisterProviderPath('^/playInvidious/$')
        def on_play_invidious(self, context, re_match):
            params = context.get_params()
                    
            if 'video_id' in params:
                xbmc.Player().play('plugin://plugin.video.invidious/?action=video&videoId=%s' % params['video_id'])
                return True
```

This is all untested btw, it usually takes a few tries to get it working right.  
When changing the code of an add-on with lots of script packages like this one, in Kodi you need to long-press the add-on icon > Information > Disable, then Enable again so it reloads all its scripts, including your updates. Otherwise the internal packages are not reloaded.

Edit: fixed a typo. Thanks to christianhaitian for the tip.

**Comment by pitsi** [comment]
@anxdpanic 
Thank you for uploading the alpha versions in the testing repo!

**Comment by Dendraspis** [comment]
I have the same issues since a few weeks after updating to Kodi v19.
Meanwhile I noticed that it helps to delete the data cache und function databases as well as the temp files via the plugin settings. It seemed, that the playback was smoother. Then I played a bit with the cache. Sadly the max size is 100MiB, which I had already set. So I changed it the min. value of 5MiB and again I flushed the databases and temp files like before. The result: Even less buffering interruptions.

Maybe coincidence or just imagination, but I would say, that the buffering issue still occurs, but noticeably less often. 
But nevertheless this buffering problem is soooo annoying...

Currently using Kodi v19.2 on Windows10 and v6.8.18.alpha2+matrix1.

**Comment by axvpast** [comment]
Looks like Google sometimes send to client broken URL with video stream. It is not problem related to your network and any Kodi buffers setup.
TL;DR
Use Odroid-C4, CoreElec 9.2.8 (last), Kodi 18.9, 4k TV (2160p@60, HDR), 1Gb WIRED Ethernet link with 100Mb Internet from ISP, which easy reach 10MiB/s speed.
I think that it is my ISP or my DNS setup, or NTP create troubles and found something interesting about this issue.
When I try to play video in folder /storage/.kodi/temp/plugin.video.youtube koid cretate mpd files.
Inside MPD files you can find video stream URL using XPath:
``
//Representation[@codecs='vp9']/BaseUrl/
``
So Kodi sometimes play this stream with dead buffering.... and you can see it using curl (checkUrl.sh):

```
LastYouTubeVideoMpd=$(ls -tr /storage/.kodi/temp/plugin.video.youtube/*.mpd | tail -n 1)

echo "Checking: ${LastYouTubeVideoMpd}"

YouTubeStreamUrlEncoded=$(cat "${LastYouTubeVideoMpd}" | grep "BaseURL" | head -n 1 | sed 's/.*<BaseURL>\([^<]*\).*/\1/' )

YouTubeStreamUrl=$( echo "${YouTubeStreamUrlEncoded}" | sed 's/\&amp;/\&/g' )

echo "Got YouTube stream URL:${YouTubeStreamUrl}"

curl -o /dev/null "${YouTubeStreamUrl}"
```
So can see curl statistics and it show that you download speed for this stream is low. But for working streams good. Also in url present interesting parameter: initcwndbps and if you got low value for this parameter you got troubles. I try to change it, but it's looks like signed and cannot be changed from our side.

So may be we just got dead/broken links from google and need ask it again till moment when we got good initcwndbps value?
Also looks like exist value for this parameter above 1Mbit/s and all below 1Mbit/s transferred as 96kBps ? Some programmer from google miss one zero?

**Comment by probonopd** [comment]
Been testing "Play with Inviduous" (as described [above](https://github.com/anxdpanic/plugin.video.youtube/issues/211#issuecomment-939445166)) for a while now and did not encounter the buffering issues with it.

**Comment by arcaine2** [comment]
I've just updated to alpha2. So far every video plays at first try, no buffering during playback so far either.

**Comment by robastala** [comment]
Was using v6.8.17 with Kodi 19.1 on Ubuntu 20.04.

Today tried: plugin.video.youtube-6.8.18.alpha2+matrix.1.zip

so far its been night and day. No buffering issues.

Thanks!

**Comment by Rialtho-the-Magnificent** [comment]
not for me. playing on pi 3b+ and OSMC as OS. Going back to the previous installment. Now it is even worse! (that is: v6.8.18.alpha2+matrix1.)

**Comment by bzarky** [comment]
Installed alpha2 on my system and it appears to have fixed the issue. So far every video is playing without issue, though some videos will just stop playing if i press any key on the keyboard. Minor issue for smooth playback.

**Comment by Nigel1992** [comment]
> Been testing "Play with Inviduous" (as described [above](https://github.com/anxdpanic/plugin.video.youtube/issues/211#issuecomment-939445166)) for a while now and did not encounter the buffering issues with it.

Thanks a lot for posting this workaround.

Working flawlessly on LibreELEC 9.2.8 (Kodi 18) on RPI 3B+

**Comment by pitsi** [comment]
I can try it on 18 and report back, provided that someone makes it a zip ready for installation because the code is too complex for me and I will probably mess something.

**Comment by RNavega** [comment]
@pitsi good to hear.  If you can wait a couple of days, I'd like to try modding the latest version of this plugin to add some scraping code, so it can stream through these Invidious proxies without the need for that external add-on.   
This Invidious service looks great, very promising.

**Comment by Rialtho-the-Magnificent** [comment]
gone back to previous installment but that makes no difference. Now no video plays oke, except one from a channel that is not using the ad revenue system I believe.

**Comment by pitsi** [comment]
@RNavega 
I can wait as long as you want. But...
I have been using plugin.video.invidious addon by lekma as an "altenative" for 6+ months now. It has some issues but they are not related to the plugin itself but to the invidious api and more specificaly to its instances. So please, use an instance that works fairly well for everyone.
Thank you in advance.

**Comment by probonopd** [comment]
Maybe a working instance could be picked automatically (randomized?) by consulting https://api.invidious.io/instances.json.

Better yet, study what they are doing (it's open source!) and implement the same strategy.

**Comment by axvpast** [comment]
Expecting invidious.io service crash after new Kodi YouTube plugin release :)
Also I think brutal solution, with a lot of parallel open TCP connections win (Google limit speed to 70bBps? Just open 10-20 connections with small ranges in parallel). Return to NetVampire download strategy :)

Simple logic - open connection, try to get 0-10MiB range. Look into speed and calculate next range offset and size for next connection, if speed for sum firs connection and next is not good - calculate offset and size for next connection... So in case if speed at the begin is good - you use only one tcp connection. If speed is not good - a lot of connections which fills buffer in parallel. So you get buffering lag only in the begin of video.

**Comment by innerspacepilot** [comment]
This seems to be a problem with outdated youtube-dl
`youtube-dl 'https://www.youtube.com/watch?v=a3oQy6v3MrA'
[youtube] a3oQy6v3MrA: Downloading webpage
[download] Destination: (HOW) I fixed the Ugly Layers on my 3D Printer-a3oQy6v3MrA.f137.mp4
[download]   0.3% of 146.70MiB at 77.57KiB/s ETA 32:11^C
ERROR: Interrupted by user
`
Comparing to fresh youtube-dl fork
`yt-dlp -N1 'https://www.youtube.com/watch?v=a3oQy6v3MrA'
[youtube] a3oQy6v3MrA: Downloading webpage
[youtube] a3oQy6v3MrA: Downloading android player API JSON
[info] a3oQy6v3MrA: Downloading 1 format(s): 248+251
[download] Resuming download at byte 4904480
[download] Destination: (HOW) I fixed the Ugly Layers on my 3D Printer [a3oQy6v3MrA].f248.webm
[download]   9.0% of 74.17MiB at  1.23MiB/s ETA 00:54^C
ERROR: Interrupted by user
`

**Comment by Nigel1992** [comment]
> > Does anyone know how to make "Play with Invidious" the default action when a video is clicked?
> 
> @probonopd
> 
> 1. When a video item is created, it uses the `/play/video_id=...` route, created in here:
>    https://github.com/anxdpanic/plugin.video.youtube/blob/36e037518bbb14a06a4928c9dd5b77baaa70c3d3/resources/lib/youtube_plugin/youtube/helper/v3.py#L54
> 2. That route is processed in here:
>    https://github.com/anxdpanic/plugin.video.youtube/blob/36e037518bbb14a06a4928c9dd5b77baaa70c3d3/resources/lib/youtube_plugin/youtube/provider.py#L683-L698
> 
> So if you want to have "Play With Invidious" as the default action, while still supporting playing in the original way, then you'll have to change that line in item 1 to be something like:
> 
> ```
>             item_uri = context.create_uri(['playInvidious'], item_params)
> ```
> 
> And then add this new "playInvidious" route to your [provider.py](https://github.com/anxdpanic/plugin.video.youtube/blob/master/resources/lib/youtube_plugin/youtube/provider.py):
> 
> ```
>         @kodion.RegisterProviderPath('^/playInvidious/$')
>         def on_play_invidious(self, context, re_match):
>             params = context.get_params()
>                     
>             if 'video_id' in params:
>                 xbmc.Player.play('plugin://plugin.video.invidious/?action=video&videoId=%s' % params['video_id'])
>                 return True
> ```
> 
> This is all untested btw, it usually takes a few tries to get it working right. When changing the code of an add-on with lots of script packages like this one, in Kodi you need to long-press the add-on icon > Information > Disable, then Enable again so it reloads all its scripts, including your updates. Otherwise the internal packages are not reloaded.

Gives a YouTube error after updating the provider file.
I probably put it in the wrong section.
Please help @RNavega  :)
My provider.py file: https://controlc.com/531e8380

Kodi 18.9 (Leia)
LibreELEC 9.2.8 (Stable)
RPi 3B+


Edit: never mind
I'll wait for your modded version of this addon 👍

**Comment by robastala** [comment]
> 
> 
> Was using v6.8.17 with Kodi 19.1 on Ubuntu 20.04.
> 
> Today tried: plugin.video.youtube-6.8.18.alpha2+matrix.1.zip
> 
> so far its been night and day. No buffering issues.
> 
> Thanks!

I spoke too soon. The buffering problem gradually came back.

**Comment by z0w13** [comment]
Feels like YouTube keeps messing with certain parameters to make playback more difficult, all my videos are buffering now too. Come on Susan

**Comment by tssajo** [comment]
> Maybe a working instance could be picked automatically (randomized?) by consulting https://api.invidious.io/instances.json.
> 
> Better yet, study what they are doing (it's open source!) and implement the same strategy.

I already looked into it last night. Invidious seems to be written in Crystal programming language??
See: https://en.wikipedia.org/wiki/Crystal_(programming_language)
Anyhow, they figured out a way to get working video stream links from Youtube without any need to decode any obfuscated data. This is achieved by using something called "Innertube API/Youtube internal API."
Their solution is based on findings mentioned here: https://github.com/TeamNewPipe/NewPipeExtractor/issues/562
This technique is implemented in this Crystal ( .cr ) file:
https://github.com/iv-org/invidious/blob/master/src/invidious/helpers/youtube_api.cr
It doesn't look to complex to implement it in Python, IMO.

**Comment by arcaine2** [comment]
> Feels like YouTube keeps messing with certain parameters to make playback more difficult, all my videos are buffering now too. Come on Susan

Yup, it's unplayable again. A single video can take 10, 20 or even more attempts to play correctly. What's intresting, the same video, from the same playlist works at first try on my Shield TV, also in Kodi with same version of the add-on.

**Comment by viper-666** [comment]
@probonopd 
I have get managed to get invidious installed on my libreelec installation and invidious works fine.
But I can't get the counter menu entry for invidious. I edited the two files as you said to do but if I go on a video in the YouTube plugin and use the counter menu there isn't a "play with invidious" entry.
Any ideas what's I make wrong...

EDIT:
I found the issue, I used the wrong "def update_play_info" line, there are two and the second is right, now everything works!!!

**Comment by pitsi** [comment]
> Maybe a working instance could be picked automatically (randomized?) by consulting https://api.invidious.io/instances.json.
> 
> Better yet, study what they are doing (it's open source!) and implement the same strategy.

The term "working" is a brave statement for the invidious api. Being a user of the addon for those 6+ months, I can assure you that even "100% health" for an instance means nothing. I have seen instances with 100% or very near to 100% health (e.g. 99 or 98) not returning any results for searches in general, not returning any results only for channels, returning results but not playing any of them, returning results and playing them with no sound and so on. Plus the wrong results in any search.

On top of that, instances come and go. A reliable instance I was using a few months ago does not even exist anymore. Luckily, the one I am now using (ytprivate.com) has been working for the last couple of months. I hope it still works though, I have not used it for a couple of weeks.

Anyway, I hope that the patch only "converts" the relevant url e.g. from 
`https://www.youtube.com/watch?v=wpWFV41U5Wk`
to
`https://yewtu.be/watch?v=wpWFV41U5Wk`
and does the searches with youtube.

p.s. I did not choose yewtu.be randomly for the example. It works via a browser, but fails in the invidious addon. And it now has 100% health as seen here, in this user friendly table.
https://redirect.invidious.io/

**Comment by notimp** [comment]
Still having the buffering issue with the alpha 2 release on the following video:
https://www.youtube.com/watch?v=7MLdehTdNhk

**Comment by pitsi** [comment]
I also have videos that buffer when playing. It seems I have to use the addon more often than once every 3-4 days.
Unlike last time that throttling was starting after 5 seconds, it now starts at 20-25 seconds in but once it is done buffering, it keeps playing until the end with no issue.

Kodi 18.9 on libreelec 9.2.6 x64, youtube 6.8.18-alpha2 and mpeg-dash enabled.

**Comment by Morgyn** [comment]
I believe youtube-dl is suffering with the same issue, they mention a fork that has it fixed: https://github.com/ytdl-org/youtube-dl/issues/29326

**Comment by S1lvr** [comment]
Can confirm same issue as everyone else, Kodi 18 on Raspbian

**Comment by urbanboy** [comment]
Same issue at Kodi Leia Libreelec RPI 3 B+.

**Comment by pitsi** [comment]
I think I made a small discovery. Despite having mpeg-dash enabled, I have set the addon to play videos at 720p.
If i set the addon to always ask me for the quality, or if I right click on a video and select "play (ask for quality)", it will pop up a dialog box with the available options for it. The top option that has "mpd" as the "filetype" is the one that uses mpeg-dash.

Right now, that option no longer shows up on most videos, but it shows up as it used to for live streams. And guess which ones play as slideshows and which ones play with no issues... 
So, could this be the culprit? Maybe youtube's api blocked the addon's ability to use the mpeg-dash for the videos, so we are back to v6.8.17 behaviour :(

---edit
@Morgyn
You should see how yt-dlp works around that throttling issue. It tricks youtube that it sends the stream to the youtube android app! Check its changelog for 2021.07.07 here
https://github.com/yt-dlp/yt-dlp/releases/tag/2021.07.07

**Comment by gatecrasher777** [comment]
> > Was using v6.8.17 with Kodi 19.1 on Ubuntu 20.04.
> > Today tried: plugin.video.youtube-6.8.18.alpha2+matrix.1.zip
> > so far its been night and day. No buffering issues.
> > Thanks!
> 
> I spoke too soon. The buffering problem gradually came back.

It gradually comes back because emulated n-parameter transforms fail on newer/different releases of YT's base.js player code.  The only fool-proof way to around it is to run the actual base.js n-transform functions...  or like with yt-dlp fool Google into thinking you are on an Android device, which seems to work for now.

Perhaps moving towards using YT's innertube api like in [ytcog](https://github.com/gatecrasher777/ytcog) with persistent sessions will give more reliable streams too. Rather than repeatedly hitting watch pages and getting different players - this is something that doesn't happen in a web user's browser. In the browser you will keep the same player base.js throughout the "session", and all subsequent info requests are serviced via the innertube api.

What Google is doing here is trying to ensure that people are using embedded code or viewing videos on their app or website, and penalizing those it thinks are circumventing their business interests.

**Comment by viper-666** [comment]
> I think I made a small discovery. Despite having mpeg-dash enabled, I have set the addon to play videos at 720p. If i set the addon to always ask me for the quality, or if I right click on a video and select "play (ask for quality)", it will pop up a dialog box with the available options for it. The top option that has "mpd" as the "filetype" is the one that uses mpeg-dash.
> 
> Right now, that option no longer shows up on most videos, but it shows up as it used to for live streams. And guess which ones play as slideshows and which ones play with no issues... So, could this be the culprit? Maybe youtube's api blocked the addon's ability to use the mpeg-dash for the videos, so we are back to v6.8.17 behaviour :(
> 
> ---edit @Morgyn You should see how yt-dlp works around that throttling issue. It tricks youtube that it sends the stream to the youtube android app! Check its changelog for 2021.07.07 here https://github.com/yt-dlp/yt-dlp/releases/tag/2021.07.07

I haven't MPEG-Dash on and always choose the video quality but have the stuttering issue.

**Comment by gatecrasher777** [comment]
Just to be clear. Streams will buffer randomly on any hardware, any code base, any language, any stream type, any platform/os (except android) where the n-parameter challenge fails.  Because that is what Google have designed it to do.

**Comment by develroo** [comment]
Yeah well seem youtube-dl itself is being throttled so anything else based on it is as well. I can take the same video and download it with yt-dlp and it is at full speed. I think some small refactoring is in order to add either other external downloaders, or include the code for it. 

It is quite frustrating *sighs*

**Comment by arcaine2** [comment]
Funny enough, i couldn't play any single video yesterday on my desktop machine with Kodi and alpha2, but everything played no problem on Shield TV with Kodi and alpha2. Same API keys in both. Today, everything plays correctly at first attempt on the desktop, no buffering, no single video i had to re-try so far.

**Comment by probonopd** [comment]
> I did not choose yewtu.be randomly for the example. It works via a browser, but fails in the invidious addon.

That particular instance has API access disabled: https://yewtu.be/api/v1/videos/

I am not suggesting that the solution should be to access Invidious API, all I am pointing out is that Invidious seems to have found a way to produce [video URLs](https://vid.puffyan.us/api/v1/videos/aqz-KE-bpKQ?fields=dashUrl,formatStreams&pretty=1) that apparently don't have the throttling issue. Ideally the same logic that is used in Invidious could be used in this plugin. (My hack above is just a temporary workaround that happens to work well in the meantime.)

> Still having the buffering issue with the alpha 2 release on the following video:
https://www.youtube.com/watch?v=7MLdehTdNhk

The video URLs for this video from https://vid.puffyan.us/api/v1/videos/7MLdehTdNhk?fields=dashUrl,formatStreams&pretty=1 seem to play well for me.

**Comment by axvpast** [comment]
Here is looks like working solution:
https://github.com/ytdl-org/youtube-dl/issues/29326#issuecomment-942387724
Sourcecode may be interesting, but program shows what exactly it sends to google to obtain proper MPD file with non broken links. Programmers, where I can replace all required 2 parameters?
```
CoreELEC:~/.kodi/addons/script.module.youtube.dl  # fgrep -rn "clientName" */*/*/*.py
lib/youtube_dl/extractor/youtube.py:260:                'clientName': 'ANDROID',
lib/youtube_dl/extractor/youtube.py:2489:                    'clientName': 'ANDROID',
lib/youtube_dl/extractor/youtube.py:2871:                    'clientName': 'ANDROID',
CoreELEC:~/.kodi/addons/script.module.youtube.dl/lib/youtube_dl/extractor  # diff -u youtube.py.orig youtube.py
--- youtube.py.orig	2021-10-13 19:40:21.000000000 +0300
+++ youtube.py	2021-10-13 19:40:03.000000000 +0300
@@ -257,8 +257,8 @@
     _DEFAULT_API_DATA = {
         'context': {
             'client': {
-                'clientName': 'WEB',
-                'clientVersion': '2.20201021.03.00',
+                'clientName': 'ANDROID',
+                'clientVersion': '16.05',
             }
         },
     }
@@ -2486,8 +2486,8 @@
         data = {
             'context': {
                 'client': {
-                    'clientName': 'WEB',
-                    'clientVersion': '2.20201021.03.00',
+                    'clientName': 'ANDROID',
+                    'clientVersion': '16.05',
                 }
             },
         }
@@ -2868,8 +2868,8 @@
         data = {
             'context': {
                 'client': {
-                    'clientName': 'WEB',
-                    'clientVersion': '2.20201021.03.00',
+                    'clientName': 'ANDROID',
+                    'clientVersion': '16.05',
                 }
             },
             'query': query,
```
It does not help. What I miss else?

**Comment by pitsi** [comment]
> That particular instance has API access disabled: https://yewtu.be/api/v1/videos/

Thank you! I had read that when the original invidio.us shut down(1) a year ago, but I could not remember which instance it was.
And here(2) is a better table of the instances and their health status.
(1) https://omar.yt/posts/stepping-away-from-open-source
(2) https://api.invidious.io/

**Comment by michel130** [comment]
I install the alpha 2 on my Kodi Lelei on android TV.
I was woking ones, the second youtube movie stopped every couple of sec and was starting to buffer.

**Comment by klebkuchen** [comment]
> Here is a workaround that works reliably around this issue (while waiting for a proper fix):
... 
> but it seems it is golden.

@probonopd You Sir, are a wizard!! Thank you!!
Finally a workaround that fixed the issue for me on my RPi3 with LibreELEC. :100:

**Comment by pitsi** [comment]
Question about the patch that is mentioned above.
As seen, it patches youtube.py of script.module.youtube.dl. But kodi's addon does not use that in order to work, so how does it benefit from that patch? I don't even have it installed!
Moreover, script.module.youtube.dl was last updated on March, 6+ months ago and 3+ months before the last release of youtube-dl which was in early June.

**Comment by axvpast** [comment]
@pitsi thanks, I think that plugin use youtube-dl and be little bit confused, because my fix does not help. Do you know where I can change "WEB" to "ANDROID" in plugin.video.youtube?
Query for MPD, after that, should looks like:
```
POST /youtubei/v1/player HTTP/1.1
Host: www.youtube.com
X-Goog-Api-Key: A.................8

{"context":{"client":{"clientName":"ANDROID","clientVersion":"16.05"}},"videoId":"Vp9I_m6znMM"}
```

**Comment by pitsi** [comment]
Sorry, I have no idea. As I have said in various issue reports, my coding skills are close to 0.
If I knew anything, I would have ported yt-dlp's fix for the throttling since day 1.

Does anyone else notice the change in the right click > play (select quality) popup that I mentioned above? Or is it just me?

**Comment by RNavega** [comment]
@axvpast you can grab the files that were updated by [pull-request 216](https://github.com/anxdpanic/plugin.video.youtube/pull/216/files), here (right-click, Save As):

- [/resources/lib/youtube_plugin/kodion/impl/xbmc/xbmc_items.py](https://github.com/anxdpanic/plugin.video.youtube/raw/521a6fdaf75d35a9db7ed72a0ac2aae472ef4cef/resources/lib/youtube_plugin/kodion/impl/xbmc/xbmc_items.py)
- [/resources/lib/youtube_plugin/youtube/helper/video_info.py](https://github.com/anxdpanic/plugin.video.youtube/raw/521a6fdaf75d35a9db7ed72a0ac2aae472ef4cef/resources/lib/youtube_plugin/youtube/helper/video_info.py)

They're changed in that way already.  
To actually use them, you can download the zip for [6.8.18~alpha2](https://github.com/anxdpanic/plugin.video.youtube/releases/tag/6.x.x-dev), extract it to some place, replace those two files in the extracted files, zip it all back as it was and then install from zip in Kodi.

**Comment by axvpast** [comment]
@RNavega thanks a lot. Just get full file video_info.py from pull request and paste into Kodi. Looks like 100% works! According links to video streams from MPD file c=ANDROID, field aitems is missing, as expected.

**Comment by kuene** [comment]
@RNavega This works on osmc. Thanks!

**Comment by develroo** [comment]
> @RNavega This works on osmc. Thanks!

~~Does it? I have 6.8.18-alpha2 installed (had to fix the bogus dependency of xmbc-python-2.2.0, OSMC uses 3.0.0) but I am still getting some buffering. It is not as bad as it was and I can pause it for 30 seconds and it will usually continue. But it is by no means fixed I feel. ~~


EDIT: Oh wait I see I have to patch manually. Sorry. Will do that now.

**Comment by pitsi** [comment]
There is a python3 build available, it is the one that has the -matrix suffix in its name, because python3 is a matrix-only feature and osmc uses kodi 19.

**Comment by develroo** [comment]
Oh, right. I did wonder. The one official from kodi is not been update, and so I did not know it was a separate build. 

Thanks.

**Comment by podkilla** [comment]
> @axvpast you can grab the files that were updated by [pull-request 216](https://github.com/anxdpanic/plugin.video.youtube/pull/216/files), here (right-click, Save As):
> 
>     * [/resources/lib/youtube_plugin/kodion/impl/xbmc/xbmc_items.py](https://github.com/anxdpanic/plugin.video.youtube/raw/521a6fdaf75d35a9db7ed72a0ac2aae472ef4cef/resources/lib/youtube_plugin/kodion/impl/xbmc/xbmc_items.py)
> 
>     * [/resources/lib/youtube_plugin/youtube/helper/video_info.py](https://github.com/anxdpanic/plugin.video.youtube/raw/521a6fdaf75d35a9db7ed72a0ac2aae472ef4cef/resources/lib/youtube_plugin/youtube/helper/video_info.py)
> 
> 
> They're changed in that way already. To actually use them, you can download the zip for [6.8.18~alpha2](https://github.com/anxdpanic/plugin.video.youtube/releases/tag/6.x.x-dev), extract it to some place, replace those two files in the extracted files, zip it all back as it was and then install from zip in Kodi.

THIS seems to work. Just installed and played a couple of videos that lagged and buffered before, they now run smoothly.

**Comment by pitsi** [comment]
> Oh, right. I did wonder. The one official from kodi is not been update, and so I did not know it was a separate build.
> 
> Thanks.

The one from the official repo will be updated when the dev thinks 6.8.18 is good enough to be released as stable.

**Comment by develroo** [comment]
OK well FWIW It seem with the two above patched files it is working well  for me. A few seconds to get Title metadata and cache but then the video start up right are no buffering. 

So thanks to @RNavega for his help. Works great on a Vero4K+

**Comment by Rialtho-the-Magnificent** [comment]
> @axvpast you can grab the files that were updated by [pull-request 216](https://github.com/anxdpanic/plugin.video.youtube/pull/216/files), here (right-click, Save As):
> 
> * [/resources/lib/youtube_plugin/kodion/impl/xbmc/xbmc_items.py](https://github.com/anxdpanic/plugin.video.youtube/raw/521a6fdaf75d35a9db7ed72a0ac2aae472ef4cef/resources/lib/youtube_plugin/kodion/impl/xbmc/xbmc_items.py)
> * [/resources/lib/youtube_plugin/youtube/helper/video_info.py](https://github.com/anxdpanic/plugin.video.youtube/raw/521a6fdaf75d35a9db7ed72a0ac2aae472ef4cef/resources/lib/youtube_plugin/youtube/helper/video_info.py)
> 
> They're changed in that way already. To actually use them, you can download the zip for [6.8.18~alpha2](https://github.com/anxdpanic/plugin.video.youtube/releases/tag/6.x.x-dev), extract it to some place, replace those two files in the extracted files, zip it all back as it was and then install from zip in Kodi.

seems to be working ok for me on pi 3b+ and osmc. If anyone needs to modified zip file, let me know.

**Comment by smutax** [comment]
> @RNavega thanks a lot. Just get full file video_info.py from pull request and paste into Kodi. Looks like 100% works! According links to video streams from MPD file c=ANDROID, field aitems is missing, as expected.

Just done the above on my pi 3a osmc kodi 18.9 and it seem to have fixed the buffering. Also I don't have "mpeg-dash" enabled which, was my understanding, needed in order for this to work. Many thanks. Just reporting.

**Comment by agolks** [comment]
great!
works on pi4, libreelec 9.2.8, kodi 18.9, too!
thanks!

**Comment by RNavega** [comment]
> Also I don't have "mpeg-dash" enabled which, was my understanding, needed in order for this to work.

@smutax it should work with adaptive formats (what you get when Mpeg DASH is on) as well as the standard formats when it's off.  
Also useful to know,  nowadays YT only has 360p and 720p as standard formats (edit: most of the time, that is). Any other sizes (either bigger or smaller) are only available if you turn on DASH in the settings. For instance,  1440p, 1080p, 240p.

> The one from the official repo will be updated when the dev thinks 6.8.18 is good enough to be released as stable.

@pitsi that's right.  The problem right now is that age-restricted videos can't be played with this experimental patch,  because it uses a new API that's obscure about how to use the login credentials.

**Comment by ChopperRob** [comment]
updated both files and it is working again, thx

**Comment by mikromcz** [comment]
Kodi 19.2 on Ubuntu here. Two files fix FINALLY worked as a charm! It is now working better than ever before. Thank you 🙏

**Comment by prometheanfire** [comment]
```sh
wget https://raw.githubusercontent.com/anxdpanic/plugin.video.youtube/521a6fdaf75d35a9db7ed72a0ac2aae472ef4cef/resources/lib/youtube_plugin/youtube/helper/video_info.py -O ~/.kodi/addons/plugin.video.youtube/resources/lib/youtu
be_plugin/youtube/helper/video_info.py
wget https://raw.githubusercontent.com/anxdpanic/plugin.video.youtube/521a6fdaf75d35a9db7ed72a0ac2aae472ef4cef/resources/lib/youtube_plugin/kodion/impl/xbmc/xbmc_items.py -O ~/.kodi/addons/plugin.video.youtube/resources/lib/you
tube_plugin/kodion/impl/xbmc/xbmc_items.py
```

this works after installing the alpha2 matrix release

**Comment by LeJimster** [comment]
> ```shell
> wget https://raw.githubusercontent.com/anxdpanic/plugin.video.youtube/521a6fdaf75d35a9db7ed72a0ac2aae472ef4cef/resources/lib/youtube_plugin/youtube/helper/video_info.py -O ~/.kodi/addons/plugin.video.youtube/resources/lib/youtu
> be_plugin/youtube/helper/video_info.py
> wget https://raw.githubusercontent.com/anxdpanic/plugin.video.youtube/521a6fdaf75d35a9db7ed72a0ac2aae472ef4cef/resources/lib/youtube_plugin/kodion/impl/xbmc/xbmc_items.py -O ~/.kodi/addons/plugin.video.youtube/resources/lib/you
> tube_plugin/kodion/impl/xbmc/xbmc_items.py
> ```
> 
> this works after installing the alpha2 matrix release

Only tried a few videos, but this works!  I've been experiencing the buffering and slow loading for ages (feels like years) to a point where every video is doing it.  I hope this fix lasts.

**Comment by Anonymous07070707** [comment]
> 
> 
> > ```shell
> > wget https://raw.githubusercontent.com/anxdpanic/plugin.video.youtube/521a6fdaf75d35a9db7ed72a0ac2aae472ef4cef/resources/lib/youtube_plugin/youtube/helper/video_info.py -O ~/.kodi/addons/plugin.video.youtube/resources/lib/youtu
> > be_plugin/youtube/helper/video_info.py
> > wget https://raw.githubusercontent.com/anxdpanic/plugin.video.youtube/521a6fdaf75d35a9db7ed72a0ac2aae472ef4cef/resources/lib/youtube_plugin/kodion/impl/xbmc/xbmc_items.py -O ~/.kodi/addons/plugin.video.youtube/resources/lib/you
> > tube_plugin/kodion/impl/xbmc/xbmc_items.py
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > this works after installing the alpha2 matrix release
> 
> Only tried a few videos, but this works! I've been experiencing the buffering and slow loading for ages (feels like years) to a point where every video is doing it. I hope this fix lasts.

I've another problem. In this fix method live streams are not playable.

**Comment by vnnegt** [comment]
@RNavega

thx, works perfectly on Nvidia Shield Kodi v18.9

**Comment by LeJimster** [comment]
> > > ```shell
> > > wget https://raw.githubusercontent.com/anxdpanic/plugin.video.youtube/521a6fdaf75d35a9db7ed72a0ac2aae472ef4cef/resources/lib/youtube_plugin/youtube/helper/video_info.py -O ~/.kodi/addons/plugin.video.youtube/resources/lib/youtu
> > > be_plugin/youtube/helper/video_info.py
> > > wget https://raw.githubusercontent.com/anxdpanic/plugin.video.youtube/521a6fdaf75d35a9db7ed72a0ac2aae472ef4cef/resources/lib/youtube_plugin/kodion/impl/xbmc/xbmc_items.py -O ~/.kodi/addons/plugin.video.youtube/resources/lib/you
> > > tube_plugin/kodion/impl/xbmc/xbmc_items.py
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > this works after installing the alpha2 matrix release
> > 
> > 
> > Only tried a few videos, but this works! I've been experiencing the buffering and slow loading for ages (feels like years) to a point where every video is doing it. I hope this fix lasts.
> 
> I've another problem. In this fix method live streams are not playable.

Yep can confirm live streams don't work.  I chose a bunch of random live streams and they all failed..  I think this part of the log file is relevant to the bug?

> 2021-10-15 01:49:52.438 T:150722    INFO <general>: [plugin.video.youtube] Running: YouTube (6.8.18+matrix.1~alpha2) on Matrix (Kodi-19.1) with Python 3.9.7
>                                                    	Path: /play/
>                                                    	Params: {'video_id': 'HNSnlQVj29E'}
> 2021-10-15 01:49:52.705 T:150722   ERROR <general>: EXCEPTION Thrown (PythonToCppException) : -->Python callback/script returned the following error<--
>                                                     - NOTE: IGNORING THIS CAN LEAD TO MEMORY LEAKS!
>                                                    Error Type: <class 'requests.exceptions.InvalidHeader'>
>                                                    Error Contents: Value for header {Referer: ('https://www.youtube.com/watch?v=HNSnlQVj29E',)} must be of type str or bytes, not <class 'tuple'>

**Comment by frankrbrts622** [comment]
Thanks! The patch fixes the problem for me (Kodi, 18.9 on 2012 macMini). This had to be a Youtube problem, as all downloaders (YT-DL, JDownloader) weren't working this week, only YT-DLP worked as usual). It was clear that the files were buffering at 40K instead of the normal speed when I looked at the info screen while trying to watch a video. Sometimes the buffer would get to 10-20% of the video file, then it would run OK until the slow speed caught up to where the video was actually playing, then it would stall. Again - many, many thanks!

**Comment by spirgiuz** [comment]
@Anonymous07070707 @LeJimster you can fix the livestream playback problem by removing the comma at the end of line 658 in video_info.py:
basically change 
`headers['Referer'] = 'https://www.youtube.com/watch?v=%s' % video_id,`
to
`headers['Referer'] = 'https://www.youtube.com/watch?v=%s' % video_id`

**Comment by viper-666** [comment]
@RNavega 
I' m so thankful, I can look Youtube since month without the buffering issue. How fast the video now starts!!!
My system RPI 3b with libreelec and patched (like you mentioned) 6.8.18~alpha2 youtube plugin.

**Comment by LeJimster** [comment]
@spirgiuz Nice, yeah that works, simple fix..   
@RNavega
Thanks to everyone for the great work!  It's nice to have the youtube plugin working properly for the first time in ages.

**Comment by Ch1llb0** [comment]
I can confirm this to be working as well! 👍🏻

**Comment by Anonymous07070707** [comment]
> 
> 
> @Anonymous07070707 @LeJimster you can fix the livestream playback problem by removing the comma at the end of line 658 in video_info.py: basically change `headers['Referer'] = 'https://www.youtube.com/watch?v=%s' % video_id,` to `headers['Referer'] = 'https://www.youtube.com/watch?v=%s' % video_id`

Thanks!

**Comment by mrmarkuz** [comment]
> 
> 
> ```shell
> wget https://raw.githubusercontent.com/anxdpanic/plugin.video.youtube/521a6fdaf75d35a9db7ed72a0ac2aae472ef4cef/resources/lib/youtube_plugin/youtube/helper/video_info.py -O ~/.kodi/addons/plugin.video.youtube/resources/lib/youtu
> be_plugin/youtube/helper/video_info.py
> wget https://raw.githubusercontent.com/anxdpanic/plugin.video.youtube/521a6fdaf75d35a9db7ed72a0ac2aae472ef4cef/resources/lib/youtube_plugin/kodion/impl/xbmc/xbmc_items.py -O ~/.kodi/addons/plugin.video.youtube/resources/lib/you
> tube_plugin/kodion/impl/xbmc/xbmc_items.py
> ```
> 
> this works after installing the alpha2 matrix release

Working here on LibreELEC 10.0.0. Thanks!

**Comment by pitsi** [comment]
Are you guys sure about this thing in line 658? This is what lines 656 to 663 say in my  `.kodi/addons/plugin.video.youtube/resources/lib/youtube_plugin/youtube/helper/video_info.py`
```
 656         headers = {
 657             'Connection': 'keep-alive',
 658             'User-Agent': 'Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/39.0.2171.3>
 659             'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8',
 660             'DNT': '1',
 661             'Accept-Encoding': 'gzip, deflate',
 662             'Accept-Language': 'en-US,en;q=0.8,de;q=0.6'
 663         }
```
6.8.18a2 on kodi 18.9, aka the python2 version.

Also, what is this garbage of an api?! I tried playing this video, a 12 HOUR long video which oddly has 4 hours of duration inside kodi, and EVERYTIME it started from 3:59:45 or so and played for 15 seconds to its end! Other videos from the same channel do not start at all.
`https://www.youtube.com/watch?v=eIwLyAOSblE`

**Comment by taleteller** [comment]
@pitsi I dont have a hand in this code, but I would assume this is an attempt to mimic some Galaxy S9 smartphone because Google is playing cat and mouse to block playback on "unapproved" apps.

**Comment by vnnegt** [comment]
> Are you guys sure about this thing in line 658? This is what lines 656 to 663 say in my `.kodi/addons/plugin.video.youtube/resources/lib/youtube_plugin/youtube/helper/video_info.py`
> 
> ```
>  656         headers = {
>  657             'Connection': 'keep-alive',
>  658             'User-Agent': 'Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/39.0.2171.3>
>  659             'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8',
>  660             'DNT': '1',
>  661             'Accept-Encoding': 'gzip, deflate',
>  662             'Accept-Language': 'en-US,en;q=0.8,de;q=0.6'
>  663         }
> ```
> 
> 6.8.18a2 on kodi 18.9, aka the python2 version.
> 
> Also, what is this garbage of an api?! I tried playing this video, a 12 HOUR long video which oddly has 4 hours of duration inside kodi, and EVERYTIME it started from 3:59:45 or so and played for 15 seconds to its end! Other videos from the same channel do not start at all. `https://www.youtube.com/watch?v=eIwLyAOSblE`

i edited the `video_info.py` file posted above directly, then inserted it in the zip file in the appropriate place (it should overwrite the existing file)

then installed the zip file

in the file above lines 656 - 662 (663 is blank) are:
```
def _load_manifest(self, url, video_id, meta_info=None, curl_headers='', playback_stats=None):
        headers = self.MOBILE_HEADERS.copy()
        headers['Referer'] = 'https://www.youtube.com/watch?v=%s' % video_id
        headers['Accept'] = '*/*'

        if playback_stats is None:
            playback_stats = {}
```

**Comment by LeJimster** [comment]
@pitsi after installing 6.8.18a2 did you download the 2 new files video_info.py & xbmc_items.py? (the download links are quoted in the post before yours). You need to edit that video_info.py.

**Comment by pitsi** [comment]
That explains a lot thank you. No, I have not made any changes after installing 6.8.18a2 (the days it was released). I will wait for a new release with the new code.

Can someone that has the files already patched check the video I mentioned above?

**Comment by LeJimster** [comment]
@pitsi yeah the same behaviour you've experienced happens when I try to play the video:
`https://www.youtube.com/watch?v=eIwLyAOSblE` 
It claims the video is only 3:59:59 long and starts at 3:59:45, plays for 15s and then stops.  I tried 3 or 4 videos from that channel and they all play fine however.

I notice it still detects the video as a **live stream** even though the stream finished hours ago.  I noticed the similar/same behaviour on a few live streams that are actually running right now, except every few seconds the play time is reset to 3:59:45 so it continues playing as long as the stream is running.  Another live stream I tried started at 0 but said it was 4 hours long.

UPDATE:  That video is now being detected as VOD rather than live and it correctly plays now as kedarz points out below.

**Comment by ilikenwf** [comment]
I just installed alpha2 and it seems to have solved the issue for me.

**Comment by kedarz** [comment]
Patched 6.8.18a2 seems that the subtitles do not work. It means, no display subtitles. Does anyone have this issue ? 
@pitsi your film works normaly, time 11:55:00 (720 mpd) 
Other films on this chanell work too

**Comment by admansymphonyck** [comment]
> I just installed alpha2 and it seems to have solved the issue for me.

For me too, but after several videos the problem returned.
I solved it after deleting the cache in addon_data, for now it's ok

**Comment by arcaine2** [comment]
This seems random, on alpha2 at least. Everything worked fine for me for the past 2 days on the desktop, not a single buffered video. Today, every video is loading slowly and buffering. Alpha2 on my Shield TV still plays everything correctly.

No problem with "play with invidious" patch from a couple posts earlier, this works each time.

**Comment by superdesi** [comment]
To all those who helped resolve this issue.....a BIG THANKS!  Now.....every video loads 3 times as fast and the buffering is gone.  In the past few days.....the add-on was unusable since every single video buffered for ages.  Stopping and starting the video has no effect and didn't help like it used to a month ago.  Thanks guys!

**Comment by Nigel1992** [comment]
Can someone can send me the zip file for Leia please ?
@RNavega @Rialtho-the-Magnificent 

My email is behind a captcha for obvious reasons.
http://scr.im/thedjs

**Comment by pitsi** [comment]
> @pitsi your film works normaly, time 11:55:00 (720 mpd) Other films on this chanell work too

Please mention your setup (just kodi version and if mpeg-dash is enabled). For me, today it behaves like the rest of the videos on this channel, it does not start at all. 
Here is all the info I can get from the log
```
2021-10-16 06:53:13.429 T:140261690197760 WARNING: Previous line repeats 5 times.
2021-10-16 06:53:13.429 T:140261690197760   ERROR: /storage/.kodi/addons/script.module.urllib3/lib/urllib3/connectionpool.py:1004: InsecureRequestWarning: Unverified HTTPS request is being made. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
                                              InsecureRequestWarning,
2021-10-16 06:53:16.146 T:140263158143104  NOTICE: VideoPlayer::OpenFile: plugin://plugin.video.youtube/play/?video_id=eIwLyAOSblE&incognito=True
2021-10-16 06:53:16.324 T:140261698590464  NOTICE: Creating InputStream
2021-10-16 06:53:16.652 T:140261698590464  NOTICE: Creating Demuxer
2021-10-16 06:53:46.766 T:140261698590464   ERROR: Open - Error, could not open file https://r3---sn-vuxbavcx-5uiel.googlevideo.com/videoplayback?expire=1634377992&ei=qExqYa3JKofV1gKb9LGYBw&ip=87.202.19.130&id=o-AJyDejb1Y6zFRNAj11qPCtFdJ7Tb4Gy8QYDP2iqNz74f&itag=22&source=youtube&requiressl=yes&mh=T7&mm=31%2C29&mn=sn-vuxbavcx-5uiel%2Csn-4g5edns6&ms=au%2Crdu&mv=m&mvi=3&pcm2cms=yes&pl=22&initcwndbps=868750&vprv=1&mime=video%2Fmp4&ns=mTdY-CKAKgyL53oH0Kz3p1UG&cnr=14&ratebypass=yes&dur=42900.050&lmt=1634318835069448&mt=1634356143&fvip=3&fexp=24001373%2C24007246&c=WEB&txp=7316222&n=d0yjjiNiqHOoz6_BB&sparams=expire%2Cei%2Cip%2Cid%2Citag%2Csource%2Crequiressl%2Cvprv%2Cmime%2Cns%2Ccnr%2Cratebypass%2Cdur%2Clmt&lsparams=mh%2Cmm%2Cmn%2Cms%2Cmv%2Cmvi%2Cpcm2cms%2Cpl%2Cinitcwndbps&lsig=AG3C_xAwRgIhAK5JfZm3Pj9wErOKykhB0ug88tiZgPeygoYW4hyxcaueAiEA3Ecy726VdTXXMPYhjnPv1smmku4WIMszRg0cd_FYlj4%3D&sig=AOq0QJ8wRQIgSlMrNl4izBwq7wUUeyHQxr5kFzkQBKAIUOZjAYTGmKYCIQDMOyAmhn-uYKWBv3GHGxBTK7JhanbTD60PmKLZa6GAIQ==
2021-10-16 06:53:46.766 T:140261698590464   ERROR: OpenDemuxStream - Error creating demuxer
2021-10-16 06:53:46.766 T:140261698590464  NOTICE: CVideoPlayer::OnExit()
2021-10-16 06:53:46.780 T:140263158143104  NOTICE: CVideoPlayer::CloseFile()
2021-10-16 06:53:46.787 T:140263158143104  NOTICE: VideoPlayer: waiting for threads to exit
2021-10-16 06:53:46.787 T:140263158143104  NOTICE: VideoPlayer: finished waiting
```
Don't ask me about script.module.urllib3. It is a dependency of neither plugin.video.youtube, nor inputstream.adaptive.

**Comment by rayanfer32** [comment]
Huge Thanks to the Folks who worked on the FIX so fast, and made the fix speed up the initial loading of video. YouTube add-on just got awesome after the patch.
The buffering is completely eradicated i feel. Never found a buffering pause after the patch.
I'm on RPI 3B+ and the video starts to play within 4 seconds after i press - play on kodi, which is a huge speed up since it took 10secs previously.

**Comment by sirjeannot** [comment]
Many thanks! 
plugin.video.youtube-6.8.18.alpha2 + the 2 patched files and now working again.
libreelec 9.2.3, kodi leia, rpi3b+

**Comment by Overbord** [comment]
Thanks so much, modified alpha2 with the 2 files, and all is working perfectly.

**Comment by Andrewrwebsteruk** [comment]
Hi Thanks, now working for me RPi2 and on RP3i., LireELEC 9.2.8 with Kodi 18.9. MPEG-DASH disabled. I couldn't figure out how to download the replacement files, I guess it's obvious once you know how. So I downloaded plugin.video.youtube-master.zip and installed that. Thanks for your hard work on fixing this issue.

**Comment by Kellett** [comment]
Alpha 2 has fixed all the buffer issues for me but thumbnails just show as the default red YouTube logo.

**Comment by ghost** [comment]
Yep, alpha2 + 2 patched files work for me on latest OSMC.

**Comment by kedarz** [comment]
> > @pitsi your film works normaly, time 11:55:00 (720 mpd) Other films on this chanell work too
> 
> Please mention your setup (just kodi version and if mpeg-dash is enabled). For me, today it behaves like the rest of the videos on this channel, it does not start at all. Here is all the info I can get from the log
> 
> ```
> 2021-10-16 06:53:13.429 T:140261690197760 WARNING: Previous line repeats 5 times.
> 2021-10-16 06:53:13.429 T:140261690197760   ERROR: /storage/.kodi/addons/script.module.urllib3/lib/urllib3/connectionpool.py:1004: InsecureRequestWarning: Unverified HTTPS request is being made. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
>                                               InsecureRequestWarning,
> 2021-10-16 06:53:16.146 T:140263158143104  NOTICE: VideoPlayer::OpenFile: plugin://plugin.video.youtube/play/?video_id=eIwLyAOSblE&incognito=True
> 2021-10-16 06:53:16.324 T:140261698590464  NOTICE: Creating InputStream
> 2021-10-16 06:53:16.652 T:140261698590464  NOTICE: Creating Demuxer
> 2021-10-16 06:53:46.766 T:140261698590464   ERROR: Open - Error, could not open file https://r3---sn-vuxbavcx-5uiel.googlevideo.com/videoplayback?expire=1634377992&ei=qExqYa3JKofV1gKb9LGYBw&ip=87.202.19.130&id=o-AJyDejb1Y6zFRNAj11qPCtFdJ7Tb4Gy8QYDP2iqNz74f&itag=22&source=youtube&requiressl=yes&mh=T7&mm=31%2C29&mn=sn-vuxbavcx-5uiel%2Csn-4g5edns6&ms=au%2Crdu&mv=m&mvi=3&pcm2cms=yes&pl=22&initcwndbps=868750&vprv=1&mime=video%2Fmp4&ns=mTdY-CKAKgyL53oH0Kz3p1UG&cnr=14&ratebypass=yes&dur=42900.050&lmt=1634318835069448&mt=1634356143&fvip=3&fexp=24001373%2C24007246&c=WEB&txp=7316222&n=d0yjjiNiqHOoz6_BB&sparams=expire%2Cei%2Cip%2Cid%2Citag%2Csource%2Crequiressl%2Cvprv%2Cmime%2Cns%2Ccnr%2Cratebypass%2Cdur%2Clmt&lsparams=mh%2Cmm%2Cmn%2Cms%2Cmv%2Cmvi%2Cpcm2cms%2Cpl%2Cinitcwndbps&lsig=AG3C_xAwRgIhAK5JfZm3Pj9wErOKykhB0ug88tiZgPeygoYW4hyxcaueAiEA3Ecy726VdTXXMPYhjnPv1smmku4WIMszRg0cd_FYlj4%3D&sig=AOq0QJ8wRQIgSlMrNl4izBwq7wUUeyHQxr5kFzkQBKAIUOZjAYTGmKYCIQDMOyAmhn-uYKWBv3GHGxBTK7JhanbTD60PmKLZa6GAIQ==
> 2021-10-16 06:53:46.766 T:140261698590464   ERROR: OpenDemuxStream - Error creating demuxer
> 2021-10-16 06:53:46.766 T:140261698590464  NOTICE: CVideoPlayer::OnExit()
> 2021-10-16 06:53:46.780 T:140263158143104  NOTICE: CVideoPlayer::CloseFile()
> 2021-10-16 06:53:46.787 T:140263158143104  NOTICE: VideoPlayer: waiting for threads to exit
> 2021-10-16 06:53:46.787 T:140263158143104  NOTICE: VideoPlayer: finished waiting
> ```
> 
> Don't ask me about script.module.urllib3. It is a dependency of neither plugin.video.youtube, nor inputstream.adaptive.

Coreelec 19.2 (Kodi 19.2) mpeg--dash enabled, quality video 720p
Could someone reproduce subtitle issue ?

**Comment by sTevefor3** [comment]
Working for me.  LibreELEC 10.0.0.  TY, anxdpanic et al!

**Comment by thonton** [comment]
> Hi Thanks, now working for me RPi2 and on RP3i., LireELEC 9.2.8 with Kodi 18.9. MPEG-DASH disabled. I couldn't figure out how to download the replacement files, I guess it's obvious once you know how. So I downloaded plugin.video.youtube-master.zip and installed that. Thanks for your hard work on fixing this issue.

Works also for me.
Think alpha3 should include also all patches.

**Comment by anxdpanic** [comment]
Posted about the new alpha in the original bug thread: https://github.com/anxdpanic/plugin.video.youtube/issues/163#issuecomment-944983426

**Comment by marcopus** [comment]
@anxdpanic it was working just fine until yesterday (alpha2+patch), now after signing in twice I only get
```
Playback failed
--------------------------------
One or more items failed to play
```
gonna check the logs..

NVM, restarting kodi did the trick

**Comment by johanburati** [comment]
Just installed [plugin.video.youtube-6.8.18.alpha3.zip](https://github.com/anxdpanic/plugin.video.youtube/releases#:~:text=plugin.video.youtube-6.8.18.alpha3.zip) this morning and it is working well. Great job guys ! 👍👍👍

**Comment by anxdpanic** [comment]
Closing this as a duplicate of #163 Please post with any follow up issues with alpha3 and buffering there.
Thanks everyone for the help and patience.
